File: B_CPU_Sched_Single_Thread/d_PS_P/ps_p.py
# Priority Scheduling // Preemptive Algorithm
# Process: [priority, AT, BT, pid]

def priority(processes):
    t = 0
    gantt = []
    table = {}
    rem_bt = {p[3]:p[2] for p in processes}

    while len(table) < len(processes):
        available = []
        for p in processes:
            if p[1] <= t and p[3] not in table: available.append(p)

        if available:
            available.sort(key=lambda x: (-x[0], x[1]))
            p = available[0]

            rem_bt[p[3]] -= 1
            t += 1
            gantt.append(p[3])
            if rem_bt[p[3]] == 0: 
                table[p[3]] = t

        else:
            gantt.append("idle")
            t += 1
    return gantt, table

if __name__ == "__main__":
    processes = [[10,0,5,'P1'], [20,1,4,'P2'], [30,2,2,'P3'], [40,4,1,'P4']]
    gantt, table = priority(processes)
    print(gantt)
    print(table)

File: B_CPU_Sched_Single_Thread/d_PS_P/ps_p.java
// Priority Scheduling // Preemptive Algorithm
// Process: [Priority, AT, BT, pid]

package B_CPU_Sched_Single_Thread.d_PS_P;
import java.util.*;
public class ps_p {

    public static Object[] priority(Object[][] processes) {
        int t = 0;
        List<String> gantt = new ArrayList<>();
        Map<String, Integer> table = new HashMap<>();

        Map<String, Integer> rem_bt = new HashMap<>();
        for (Object[] p: processes) rem_bt.put((String) p[3], (Integer) p[2]);

        while (table.size() < processes.length) {
            List<Object[]> available = new ArrayList<>();

            for (Object[] p: processes) {
                if ((Integer) p[1] <= t && !table.containsKey((String) p[3])) available.add(p);
            }

            if (available.size() == 0) {
                t++;
                gantt.add("idle");
            }
            else {
                available.sort((a,b) -> {
                    int diff = (Integer) b[0] - (Integer) a[0];
                    if (diff != 0) return diff;
                    return (Integer) a[1] - (Integer) b[1];
                });
                Object[] p = available.get(0);
                
                t++;
                String pid = (String) p[3];
                gantt.add(pid);                
                rem_bt.put(pid, rem_bt.get(pid)-1);
                if (rem_bt.get(pid) == 0) table.put(pid, t);
            }
        }
        return new Object[]{gantt, table};
    }

    public static void main(String[] args) {
        Object[] res = priority(new Object[][]{{10,0,5,"P1"}, {20,1,4,"P2"}, {30,2,2,"P3"}, {40,4,1,"P4"}});
        System.out.println(res[0].toString());
        System.out.println(res[1].toString());   
    }
}
